Mathlib/Analysis/Fourier/ZMod.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.Analysis.SpecialFunctions.Complex.Circle
+import Mathlib.Analysis.Fourier.FourierTransform
+import Mathlib.NumberTheory.DirichletCharacter.GaussSum
+
+/-!
+# Fourier theory on `ZMod N`
+
+Basic definitions and properties of the discrete Fourier transform for functions on `ZMod N`.
+
+### Main definitions and results
+
+* `ZMod.stdAddChar`: the additive character of `ZMod N` sending `1` to `exp (2 * π * I / N)`.
+* `ZMod.discreteFourierTransform`: the Fourier transform with respect to this add char. The local
+  notation `𝓕`, scoped in namespace `ZMod`, is available for this.
+* `DirichletCharacter.fourierTransform_eq_inv_mul_gaussSum`: the discrete Fourier transform of a
+  primitive Dirichlet character `χ` is a Gauss sum times `χ⁻¹`.
+-/
+
+open scoped Real
+
+namespace ZMod
+
+/-- The `AddMonoidHom` from `ZMod N` to `ℝ / ℤ` sending `j mod N` to `j / N mod 1`. -/
+noncomputable def toAddCircle {N : ℕ+} : ZMod N →+ UnitAddCircle :=
+  lift N ⟨AddMonoidHom.mk' (fun j ↦ ↑(j / N : ℝ)) (by simp [add_div]), by simp⟩
+
+lemma toAddCircle_coe {N : ℕ+} (j : ℤ) :
+    toAddCircle (j : ZMod N) = ↑(j / N : ℝ) := by
+  simp [toAddCircle]
+
+lemma toAddCircle_apply {N : ℕ+} (j : ZMod N) :
+    toAddCircle j = ↑(j.val / N : ℝ) := by
+  conv_lhs => rw [show j = (val j : ℤ) by simp, toAddCircle_coe]
+  simp only [natCast_val, intCast_cast]
+
+lemma toAddCircle_injective (N : ℕ+) : Function.Injective (toAddCircle : ZMod N → _) := by
+  intro x y hxy
+  have : (0 : ℝ) < N := Nat.cast_pos.mpr N.pos
+  rwa [toAddCircle_apply, toAddCircle_apply, AddCircle.coe_eq_coe_iff_of_mem_Ico
+    (hp := Real.fact_zero_lt_one) (a := 0), div_left_inj' this.ne', Nat.cast_inj,
+    (val_injective N).eq_iff] at hxy <;>
+  exact ⟨by positivity, by simpa only [zero_add, div_lt_one this, Nat.cast_lt] using val_lt _⟩
+
+/-- The additive character from `ZMod N` to the unit circle in `ℂ`, sending `j mod N` to
+`exp (2 * π * I * j / N)`. -/
+noncomputable def toCircle {N : ℕ+} : AddChar (ZMod N) circle where
+  toFun := fun j ↦ (toAddCircle j).toCircle
+  map_add_mul' a b := by simp_rw [map_add, AddCircle.toCircle_add]
+  map_zero_one' := by simp_rw [map_zero, AddCircle.toCircle, ← QuotientAddGroup.mk_zero,
+    Function.Periodic.lift_coe, mul_zero, expMapCircle_zero]
```

Comment:
```suggestion
noncomputable def toCircle {N : ℕ+} : AddChar (ZMod N) circle :=
  AddChar.toAddMonoidHomEquiv.symm toAddCircle 
```

Reply:
This does not work. 

Firstly `AddChar.toAddMonoidHomEquiv` has some arguments; you probably meant
`(AddChar.toAddMonoidHomEquiv _ _).symm toAddCircle`. But that just exposes the more serious problem that `UnitAddCircle` is not defined as the additivization of `circle`.

Reply:
First problem is solved in #13579. Second problem, okay sure